_posts/2017-03-16-Rust-1.16.md

```diff
@@ -0,0 +1,317 @@
+12 | |     }
+   | |_____^ ...ending here
+   |
+help: consider using an explicit lifetime parameter as shown: fn from_str(s: &'a str) -> Result<Name, ()>
+  --> .\foo.rs:10:5
+   |
+10 |       fn from_str(s: &str) -> Result<Name, ()> {
+   |  _____^ starting here...
+11 | |         Ok(Name { name: s })
+12 | |     }
+   | |_____^ ...ending here
+```
+
+The compiler explains the issue, and gives a helpful suggestion. So let's try it: modify the code to add in
+the `'a`, and compile again:
+
+```bash
+> rustc +1.15.1 .\foo.rs --crate-type=lib
+error[E0308]: method not compatible with trait
+  --> .\foo.rs:10:5
+   |
+10 |       fn from_str(s: &'a str) -> Result<Name, ()> {
+   |  _____^ starting here...
+11 | |         Ok(Name { name: s })
+12 | |     }
+   | |_____^ ...ending here: lifetime mismatch
+   |
+<snip>
+help: consider using an explicit lifetime parameter as shown: fn from_str(s: &'a str) -> Result<Name<'a>, ()>
+  --> .\foo.rs:10:5
+   |
+10 |       fn from_str(s: &'a str) -> Result<Name, ()> {
+   |  _____^ starting here...
+11 | |         Ok(Name { name: s })
+12 | |     }
+   | |_____^ ...ending here
+```
+
+It still doesn't work. That help message was not actually helpful. It does suggest adding another
+lifetime, this time on `Name`. If we do that...
+
+```bash
+> rustc +1.15.1 .\foo.rs --crate-type=lib
+<snip>
+help: consider using an explicit lifetime parameter as shown: fn from_str(s: &'a str) -> Result<Name<'a>, ()>
+  --> .\foo.rs:10:5
+```
+
+... that's what we already have, compiler!
+
+This diagnostic was well-intentioned, but when it's wrong, it was *very* wrong, as you can see here. Sometimes
+it wouldn't even suggest valid Rust syntax! Furthermore, more advanced Rust users didn't really need the
+suggestion, but new Rustaceans would take them to heart, and then be led down this bad path. As such, we decided
+that for now, [we should remove the help message entirely]. We may bring it back in the future, but only if we can
+limit false positives. 
+
+[we should remove the help message entirely]: https://github.com/rust-lang/rust/pull/37057
+
+In other diagnostic changes, previous versions of Rust would helpfully attempt to suggest fixes for typos:
+
+```rust
+let foo = 5;
+
+println!("{}", ffo);
+```
+
+Would give this error:
+
+```text
+error[E0425]: cannot find value `ffo` in this scope
+ --> foo.rs:4:20
+  |
+4 |     println!("{}", ffo);
+  |                    ^^^ did you mean `foo`?
+```
+
+However, this would only happen in certain circumstances: sometimes in local variables, and for fields in
+structs. [This now happens nearly everywhere]. When combined with [some other related improvements], this
+results in a significant improvement in these sorts of diagnostics.
+
+[This now happens nearly everywhere]: https://github.com/rust-lang/rust/pull/38927
+[some other related improvements]: https://github.com/rust-lang/rust/pull/38154
+
+Finally, `Self` can now appear [in impl headers] and [in struct expressions].
```

Comment:
`Self` in impl headers rarely works in 1.16
The final piece was landed in https://github.com/rust-lang/rust/pull/40008, so it will be fully available only in 1.17